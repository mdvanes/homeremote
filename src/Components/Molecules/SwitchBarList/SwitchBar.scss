.card {
    margin-bottom: 0.2em;

    > div {
        padding-bottom: 0 !important;
    }
}

.button-group {
    display: flex;
    padding: 0 !important;
    align-items: center; /* flex on vertical axis (perpendicular to flex direction) */

    .label {
        flex: 1;
        text-align: center;
    }

    button {
        background-color: transparent;
        border: none;
        cursor: pointer;
        padding: 0.7em 1.2em;
    }

    .dummy {
        height: 24px;
        width: 24px;
    }
}
