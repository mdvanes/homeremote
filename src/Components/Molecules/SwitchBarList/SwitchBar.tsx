import React, { FC, ReactElement } from "react";
import { Icon, Card, CardContent } from "@material-ui/core";
import useStyles from "./SwitchBar.styles";

const getIconElem = (icon: string | false): ReactElement => {
    return icon ? (
        <Icon style={{ fontSize: "250%" }} className="material-icons">
            {icon}
        </Icon>
    ) : (
        <i></i>
    );
};

type Props = {
    leftButton: ReactElement;
    rightButton: ReactElement;
    icon: string | false;
    label: string;
    labelAction: (() => void) | false;
};

/**
 * Bar with a left and right button to switch states
 * @param leftButton
 * @param rightButton
 * @param iconElem
 * @param label
 * @returns {*}
 * @constructor
 */
const SwitchBar: FC<Props> = ({
    leftButton,
    rightButton,
    icon,
    label,
    labelAction,
}) => {
    const classes = useStyles();
    return (
        <Card className={classes.card}>
            <CardContent className={classes.buttonGroup}>
                {leftButton}
                <span className={classes.label}>
                    {getIconElem(icon)}
                    {labelAction ? (
                        <button onClick={labelAction}>{label}</button>
                    ) : (
                        label
                    )}
                </span>
                {rightButton}
            </CardContent>
        </Card>
    );
};

export default SwitchBar;
