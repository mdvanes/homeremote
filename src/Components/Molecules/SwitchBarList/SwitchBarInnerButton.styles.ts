import { makeStyles, Theme, createStyles, StyleRules } from "@material-ui/core";

const useStyles = makeStyles(
    (theme: Theme): StyleRules =>
        createStyles({
            root: {
                backgroundColor: (isActive: any): string =>
                    isActive
                        ? `${theme.palette.primary.light} !important`
                        : "transparent",
            },
        })
);

export default useStyles;
