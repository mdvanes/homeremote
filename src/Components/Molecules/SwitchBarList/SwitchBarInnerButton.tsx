// eslint-disable-next-line no-unused-vars
import React from 'react';
import { Icon } from '@material-ui/core';
import { deepPurple } from '@material-ui/core/colors';

const getActiveStyle = (isActive: any) => ({
    backgroundColor: isActive ? deepPurple[500] : 'transparent'
});

const SwitchBarInnerButton = ({
    isReadOnly,
    clickAction,
    icon,
    isActive
}: any) => {
    return (
        <button
            onClick={isReadOnly ? null : clickAction}
            style={getActiveStyle(isActive)}
        >
            {isReadOnly ? (
                <div className="dummy"></div>
            ) : (
                <Icon hoverColor={deepPurple[900]} className="material-icons">
                    {icon}
                </Icon>
            )}
        </button>
    );
};

export default SwitchBarInnerButton;
