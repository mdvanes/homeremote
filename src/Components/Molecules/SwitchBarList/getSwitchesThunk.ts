import { createAsyncThunk } from "@reduxjs/toolkit";
import fetchToJson from "../../../fetchToJson";

// TODO deduplicate this fetch to util from appStatusSlice, switchBarListSlice, authenticationSlice

// Domoticz Device Switch
export type DSwitch = {
    idx: string;
    name: string;
    type: string;
    status: string;
    dimLevel: number | null;
    readOnly: boolean;
    children: DSwitch[] | false;
};

interface FetchReturned {
    switches: DSwitch[];
}

// Extracted from switchBarListSlice to be able to mock it
export const getSwitches = createAsyncThunk<FetchReturned>(
    `switchesList/getSwitches`,
    async () => fetchToJson<FetchReturned>("/api/switches")
);
