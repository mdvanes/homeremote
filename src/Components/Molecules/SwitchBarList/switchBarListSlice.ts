import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';

// Alternatively use process.env.NODE_ENV that is automatically set to development or production
const getRootUrl = (): string =>
    process.env.REACT_APP_STAGE === 'development'
        ? 'http://localhost:3001'
        : '';

export const getSwitches = createAsyncThunk(
    `switchesList/getSwitches`,
    async (_, { rejectWithValue }) => {
        try {
            const response = await fetch(`${getRootUrl()}/api/switches`, {
                credentials: 'same-origin',
                method: 'GET',
                headers: {
                    Accept: 'application/json',
                    'Content-Type': 'application/json',
                    Authorization: `Bearer ${localStorage.getItem('token')}`
                }
            });
            const json = await response.json();
            // TODO clean up
            if (json.error) {
                throw new Error(json.error);
            }
            console.log(json);
            return json;
        } catch (err) {
            console.error(err);
            return rejectWithValue([]);
        }
    }
);

export interface SwitchBarListState {
    switches: any[];
    isLoading: boolean;
}

const initialState: SwitchBarListState = { switches: [], isLoading: false };

const switchBarListSlice = createSlice({
    name: 'switchesList',
    initialState,
    reducers: {},
    extraReducers: {
        [getSwitches.fulfilled.toString()]: (state, { payload }) => {
            console.log(payload);
            state.isLoading = false;
            state.switches = payload.switches;
        },
        [getSwitches.pending.toString()]: (state, { payload }) => {
            state.isLoading = true;
            console.log('pending');
        },
        [getSwitches.rejected.toString()]: (state, { error }) => {
            state.isLoading = false;
            state.switches = [];
            console.log('error', error); // TODO dispatch to logger slice? see SwitchBarListContainer
        }
    }
});

// export const { } = switchBarListSlice.actions;

export default switchBarListSlice.reducer;
