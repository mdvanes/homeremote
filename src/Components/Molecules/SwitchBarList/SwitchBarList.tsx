import React, { FC, Fragment, ReactElement, useEffect } from "react";
import SwitchBar from "./SwitchBar";
import SwitchBarInnerButton from "./SwitchBarInnerButton";
import { useDispatch, useSelector } from "react-redux";
import {
    getSwitches,
    sendSwitchState,
    SwitchBarListState,
    toggleExpandScene,
    DSwitch,
} from "./switchBarListSlice";
import { RootState } from "../../../Reducers";
import { logError } from "../LogCard/logSlice";

// TODO improve type
const SELECTOR_STATES: Record<number, string> = {
    0: "disconnected",
    10: "disarmed",
    20: "partarmed",
    30: "armed",
};

// TODO stronger typing, return type can be only certain strings
// Type is switchscene or switchlight
const getType = (type: string): string => {
    switch (type) {
        case "Group":
            return "switchscene";
        case "Selector":
            return "selector";
        default:
            return "switchlight";
    }
};

const getLeftIcon = (label: string): string =>
    label === "Blinds" ? "arrow_drop_down" : "radio_button_checked";

const getRightIcon = (label: string): string =>
    label === "Blinds" ? "arrow_drop_up" : "radio_button_unchecked";

const getNameAndChildren = (
    name: string,
    children: DSwitch[] | false
): string => (children ? `${name} â¯†` : name);

const getLabel = (
    name: string,
    dimLevel: number | null,
    type: string
): string => {
    if (type === "selector" && dimLevel !== null) {
        return `${name}: ${SELECTOR_STATES[dimLevel]}`;
    }
    return dimLevel !== null ? `${name} (${dimLevel}%)` : name;
};

type SendSomeState = (id: string, type: string) => void;

const mapSwitchToSwitchBar = (
    { idx, name, type, dimLevel, readOnly, status, children }: DSwitch,
    sendOn: SendSomeState,
    sendOff: SendSomeState,
    labelAction: (() => void) | false
): ReactElement => (
    <SwitchBar
        key={`switch-${idx}`}
        icon={false}
        label={getLabel(
            getNameAndChildren(name, children),
            dimLevel,
            getType(type)
        )}
        labelAction={labelAction}
        leftButton={
            <SwitchBarInnerButton
                isReadOnly={readOnly}
                clickAction={(): void => sendOn(idx, getType(type))}
                icon={getLeftIcon(name)}
                isActive={status === "On"}
            />
        }
        rightButton={
            <SwitchBarInnerButton
                isReadOnly={readOnly}
                clickAction={(): void => sendOff(idx, getType(type))}
                icon={getRightIcon(name)}
                isActive={status === "Off"}
            />
        }
    />
);

const getLabelAction = (
    hasChildren: boolean,
    cbWithChildren: () => void,
    cbWithoutChildren: () => void
): (() => void) => (hasChildren ? cbWithChildren : cbWithoutChildren);

const SwitchBarList: FC = () => {
    const dispatch = useDispatch();
    const switches = useSelector<RootState, SwitchBarListState["switches"]>(
        (state: RootState) => state.switchesList.switches
    );
    const isLoading = useSelector<RootState, SwitchBarListState["isLoading"]>(
        (state: RootState) => state.switchesList.isLoading
    );
    const errorMessage = useSelector<RootState, SwitchBarListState["error"]>(
        (state: RootState) => state.switchesList.error
    );
    const expandedScenes = useSelector<
        RootState,
        SwitchBarListState["expanded"]
    >((state: RootState) => state.switchesList.expanded);

    const sendOn = (id: string, type: string): void => {
        dispatch(sendSwitchState({ id, type, state: "on" }));
    };

    const sendOff = (id: string, type: string): void => {
        dispatch(sendSwitchState({ id, type, state: "off" }));
    };

    useEffect(() => {
        dispatch(getSwitches());
    }, [dispatch]);
    useEffect(() => {
        dispatch(logError(errorMessage));
    }, [errorMessage, dispatch]);
    const switchBars = switches.map((dSwitch: DSwitch) => {
        const hasChildren = Boolean(dSwitch.children);
        const labelAction = getLabelAction(
            hasChildren,
            () => {
                dispatch(toggleExpandScene({ sceneIdx: dSwitch.idx }));
                dispatch(getSwitches());
            },
            getSwitches
        );
        const showChildren = expandedScenes.includes(dSwitch.idx);
        return (
            <Fragment key={`frag-${dSwitch.idx}`}>
                {mapSwitchToSwitchBar(dSwitch, sendOn, sendOff, labelAction)}
                {hasChildren && showChildren ? (
                    <div style={{ padding: "0.5em" }}>
                        {dSwitch.children &&
                            dSwitch.children.map((switchChild: DSwitch) =>
                                mapSwitchToSwitchBar(
                                    switchChild,
                                    sendOn,
                                    sendOff,
                                    false
                                )
                            )}
                    </div>
                ) : null}
            </Fragment>
        );
    });
    return (
        <Fragment>
            {switchBars}
            {isLoading && "loading..."}
        </Fragment>
    );
};

export default SwitchBarList;
