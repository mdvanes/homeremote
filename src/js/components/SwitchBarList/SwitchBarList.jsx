// eslint-disable-next-line no-unused-vars
import React, {Fragment} from 'react';
import SwitchBar from './SwitchBar';
import SwitchBarInnerButton from './SwitchBarInnerButton';

const SELECTOR_STATES = {
  0: 'disconnected',
  10: 'disarmed',
  20: 'partarmed',
  30: 'armed'
};

// Type is switchscene or switchlight
const getType = type => {
  switch (type) {
    case 'Group':
      return 'switchscene';
    case 'Selector':
      return 'selector';
    default:
      return 'switchlight';
  }
};

const getLeftIcon = label => label === 'Blinds' ? 'arrow_drop_down' : 'radio_button_checked';

const getRightIcon = label => label === 'Blinds' ? 'arrow_drop_up' : 'radio_button_unchecked';

const getNameAndChildren = (name, children) => children ? `${name} (${children.length})` : name;

const getLabel = ( name, dimLevel, type ) => {
  if ( type === 'selector' ) {
    return `${name}: ${SELECTOR_STATES[dimLevel]}`;
  }
  return dimLevel !== null ? `${name} (${dimLevel}%)` : name;
};

const mapSwitchToSwitchBar = ({idx, name, type, dimLevel, readOnly, status, children}, sendOn, sendOff, labelAction) => (<SwitchBar
  key={`switch-${idx}`}
  label={getLabel(getNameAndChildren(name, children), dimLevel, getType(type))}
  labelAction={labelAction}
  leftButton={<SwitchBarInnerButton
    isReadOnly={readOnly}
    clickAction={() => sendOn(idx, getType(type))}
    icon={getLeftIcon(name)}
    isActive={status === 'On'}
  />}
  rightButton={<SwitchBarInnerButton
    isReadOnly={readOnly}
    clickAction={() => sendOff(idx, getType(type))}
    icon={getRightIcon(name)}
    isActive={status === 'Off'}
  />}
/>);

const getLabelAction = (hasChildren, cb) => hasChildren ? cb : false;

const SwitchBarList = ({ switches, expandedScenes, sendOn, sendOff, toggleExpandScene }) => {
  const switchBars = switches
    .map((dSwitch) => {
      const hasChildren = dSwitch.children;
      const labelAction = getLabelAction(hasChildren, () => { toggleExpandScene(dSwitch.idx) } );
      const showChildren = expandedScenes.includes(dSwitch.idx);
      return (<Fragment key={`frag-${dSwitch.idx}`}>
        {mapSwitchToSwitchBar(dSwitch, sendOn, sendOff, labelAction)}
        {hasChildren && showChildren ? (
          <div style={{'padding': '0.5em'}}>
            {hasChildren.map(switchChild => mapSwitchToSwitchBar(switchChild, sendOn, sendOff, false))}
          </div>) : null}
      </Fragment>);
    });
  return (
    <Fragment>
      {switchBars}
    </Fragment>
  );
};

// TODO This shim can be removed if the useEffect hook can call this.props.getSwitches from the SwitchBarList FC, but that requires React >= 16.8
class SwitchBarListShim extends React.Component {
  constructor(props) {
    super(props);
    this.props.getSwitches();
  }

  render() {
    return (<SwitchBarList {...this.props} />);
  }
}

export default SwitchBarListShim;
