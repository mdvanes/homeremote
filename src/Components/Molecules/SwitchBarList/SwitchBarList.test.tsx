import React from "react";
import * as ReactRedux from "react-redux";
import { render, fireEvent, waitFor, screen } from "@testing-library/react";
import SwitchBarList from "./SwitchBarList";
import { RootState } from "../../../Reducers";

describe("SwitchBarList", () => {
    const mockDispatch = jest.fn();
    it("shows a normal light switch", () => {
        jest.spyOn(ReactRedux, "useSelector").mockReset();
        jest.spyOn(ReactRedux, "useSelector").mockImplementation(fn => {
            const mockRootState: RootState = {
                switchesList: {
                    isLoading: false,
                    error: false,
                    switches: [
                        {
                            idx: "3",
                            type: "Light/Switch",
                            name: "My Normal Light Switch",
                            status: "On",
                            dimLevel: null,
                            readOnly: false,
                            children: false,
                        },
                    ],
                    expanded: [],
                },
            };
            return fn(mockRootState);
        });
        jest.spyOn(ReactRedux, "useDispatch").mockReturnValue(mockDispatch);
        const { getByText } = render(<SwitchBarList />);
        expect(getByText("My Normal Light Switch")).toBeInTheDocument();
    });

    // TODO fix
    // it("?", () => {
    //     jest.spyOn(ReactRedux, "useSelector").mockReset();
    //     jest.spyOn(ReactRedux, "useSelector").mockImplementation(fn => {
    //         const mockRootState: RootState = {
    //             switchesList: {
    //                 isLoading: false,
    //                 error: false,
    //                 switches: [
    //                     {
    //                         idx: "3",
    //                         type: "Light/Switch",
    //                         name: "My Normal Light Switch",
    //                         status: "On",
    //                         dimLevel: null,
    //                         readOnly: false,
    //                         children: false,
    //                     },
    //                 ],
    //                 expanded: [],
    //             },
    //         };
    //         return fn(mockRootState);
    //     });
    //     jest.spyOn(ReactRedux, "useDispatch").mockReturnValue(mockDispatch);
    //     const { getByText, debug, baseElement } = render(<SwitchBarList />);
    //     // debug(baseElement)
    //     // TODO fix MUI class ids (codestar repo?)
    //     const but = baseElement.querySelector("button.makeStyles-root-78");
    //     mockDispatch.mockReset();
    //     fireEvent.click(but);
    //     debug(baseElement);
    //     expect(getByText("My Normal Light Switch")).toBeInTheDocument();
    //     expect(mockDispatch).toHaveBeenCalledWith({});
    // });

    it("shows the loading state", () => {
        jest.spyOn(ReactRedux, "useSelector").mockReset();
        jest.spyOn(ReactRedux, "useSelector").mockImplementation(fn => {
            const mockRootState: RootState = {
                switchesList: {
                    isLoading: true,
                    error: false,
                    switches: [],
                    expanded: [],
                },
            };
            return fn(mockRootState);
        });
        jest.spyOn(ReactRedux, "useDispatch").mockReturnValue(mockDispatch);
        const { getByText } = render(<SwitchBarList />);
        expect(getByText("loading...")).toBeInTheDocument();
    });
});
