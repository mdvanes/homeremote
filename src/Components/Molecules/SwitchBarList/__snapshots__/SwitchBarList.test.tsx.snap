// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwitchBarList shows a blinds switch 1`] = `
<div
  class="MuiCardContent-root makeStyles-buttonGroup"
  data-testid="switchbar-Blinds"
>
  <button
    class="makeStyles-root"
  >
    <span
      aria-hidden="true"
      class="material-icons MuiIcon-root material-icons"
    >
      arrow_drop_down
    </span>
  </button>
  <span
    class="makeStyles-label"
  >
    <i />
    <button>
      Blinds
    </button>
  </span>
  <button
    class="makeStyles-root"
  >
    <span
      aria-hidden="true"
      class="material-icons MuiIcon-root material-icons"
    >
      arrow_drop_up
    </span>
  </button>
</div>
`;

exports[`SwitchBarList shows a normal light switch 1`] = `
<div
  class="MuiCardContent-root makeStyles-buttonGroup"
  data-testid="switchbar-My Normal Light Switch"
>
  <button
    class="makeStyles-active"
  >
    <span
      aria-hidden="true"
      class="material-icons MuiIcon-root material-icons"
    >
      radio_button_checked
    </span>
  </button>
  <span
    class="makeStyles-label"
  >
    <i />
    <button>
      My Normal Light Switch
    </button>
  </span>
  <button
    class="makeStyles-root"
  >
    <span
      aria-hidden="true"
      class="material-icons MuiIcon-root material-icons"
    >
      radio_button_unchecked
    </span>
  </button>
</div>
`;

exports[`SwitchBarList shows a selector switch 1`] = `
<div
  class="MuiCardContent-root makeStyles-buttonGroup"
  data-testid="switchbar-My Selector Switch: armed"
>
  <button
    class="makeStyles-root"
  >
    <div
      class="dummy"
    />
  </button>
  <span
    class="makeStyles-label"
  >
    <i />
    <button>
      My Selector Switch: armed
    </button>
  </span>
  <button
    class="makeStyles-root"
  >
    <div
      class="dummy"
    />
  </button>
</div>
`;
